### PR TITLE
guard against deposit txs

### PR DESCRIPTION
**Pull request by AnkushinDaniil**
Now we prevent deposit txs from being included via the mempool, such as the BroadcastTx Comet API method. They should only be included via the Engine API.
[fixes](https://www.notion.so/nethermind/Make-sure-deposit-txs-are-not-executed-via-mempool-d09be4ccf95845948679031b42708d04?pvs=4)

## Summary by CodeRabbit

- **New Features**
	- Enhanced transaction validation logic to prevent deposit transactions from being enqueued in the mempool.
	- Introduced a new function for adapting Cosmos deposit transactions to Ethereum transactions.

- **Bug Fixes**
	- Improved error handling and clarity when adapting transactions, ensuring better user feedback.

- **Tests**
	- Added new test cases to validate the handling of deposit transactions in the mempool.
	- Refactored tests for clearer transaction adaptation processes and improved maintainability.

### CODE REVIEW

**Comment by AnkushinDaniil** [comment]
@joshklop 
review request

File: mempool/mempool.go [hunk 1]
```diff
@@ -36,6 +37,9 @@ func (p *Pool) Enqueue(userTxn comettypes.Tx) error {
 	// NOTE: we should do reads and writes on the same view. Right now they occur on separate views.
 	// Unfortunately, comet's DB interface doesn't support it.
 	// Moving to a different DB interface is left for future work.
+	if txs, _ := rolluptypes.AdaptCosmosDepositTxToEthTx(userTxn); len(txs) > 0 {
```
Comment (joshklop) [inline mempool/mempool.go:40]: Right now, it appears like the comment applies to these lines. Let's leave a space in between, with a comment right above your changes explaining the reasoning (to disallow deposit txs in the mempool).

**Review by joshklop — CHANGES_REQUESTED** [review]

File: mempool/mempool.go [hunk 2]
```diff
@@ -37,6 +38,14 @@ func (p *Pool) Enqueue(userTxn comettypes.Tx) error {
 	// Unfortunately, comet's DB interface doesn't support it.
 	// Moving to a different DB interface is left for future work.
 
+	// Attempt to adapt the Cosmos transaction to an Ethereum deposit transaction.
+	// If the adaptation results in one or more transactions, it indicates that the
+	// user transaction is a deposit transaction, which is not allowed in the pool.
+	if txs, _ := rolluptypes.AdaptCosmosDepositTxToEthTx(userTxn); len(txs) > 0 {
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go:44]: No, it doesn't work that way. In the case of `err == nil` when I run `make test`
```
--- FAIL: TestBroadcastTx (0.01s)
    comet_test.go:128: 
                Error Trace:    /Users/daniilankusin/monomer/comet/comet_test.go:128
                Error:          Received unexpected error:
                                enqueue in mempool: deposit txs are not allowed in the pool
                Test:           TestBroadcastTx
FAIL
```

[the remark this answers is not in the document]
Comment (joshklop) [reply mempool/mempool.go:44]: Hmmm, why does the test fail?

[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go:44]: Because of the functionality of the adapter in [here](https://github.com/AnkushinDaniil/monomer/blob/fb07012e7740b192fe4d2a57293afc904d90c231/x/rollup/types/adapters.go#L113-L115). It calls break, not an error. If I change it to return an error, the behaviour of the function is false. I can implement another function, but I don't think it's necessary. It would just be a wrap or very similar function.

[the remark this answers is not in the document]
Comment (joshklop) [reply mempool/mempool.go:44]: We should change the `break` to return an error.

[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go:44]: Yes, I tried that. In case of an error instead of break we fail other tests:
```
--- FAIL: TestBuild (0.02s)
    --- FAIL: TestBuild/txs_in_inclusion_list (0.01s)
        builder_test.go:135: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:135
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestBuild/txs_in_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool (0.00s)
        builder_test.go:91: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:91
                Error:          Received unexpected error:
                                deposit txs are not allowed in the pool
                Test:           TestBuild/txs_in_mempool
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list (0.00s)
        builder_test.go:91: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:91
                Error:          Received unexpected error:
                                deposit txs are not allowed in the pool
                Test:           TestBuild/txs_in_mempool_and_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool (0.00s)
        builder_test.go:91: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:91
                Error:          Received unexpected error:
                                deposit txs are not allowed in the pool
                Test:           TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool
--- FAIL: TestRollback (0.01s)
    builder_test.go:252: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:252
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestRollback
FAIL
FAIL    github.com/polymerdao/monomer/builder   1.042s
--- FAIL: TestBroadcastTx (0.00s)
    comet_test.go:128: 
                Error Trace:    /Users/daniilankusin/monomer/comet/comet_test.go:128
                Error:          Received unexpected error:
                                enqueue in mempool: deposit txs are not allowed in the pool
                Test:           TestBroadcastTx
```

[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go:44]: I will try to fix the builder tests because I support your point of view, I don't feel that we can leave `break` in there.

[the remark this answers is not in the document]
Comment (joshklop) [reply mempool/mempool.go:44]: We can fix the tests using the transaction helpers in `testutils` I think.

[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go:44]: Yes, I tried that, it results in this:
```
--- FAIL: TestBuild (0.03s)
    --- FAIL: TestBuild/txs_in_inclusion_list (0.00s)
        builder_test.go:135: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:135
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestBuild/txs_in_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool (0.01s)
        builder_test.go:135: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:135
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestBuild/txs_in_mempool
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list (0.00s)
        builder_test.go:135: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:135
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestBuild/txs_in_mempool_and_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool (0.00s)
        builder_test.go:135: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:135
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool
--- FAIL: TestRollback (0.01s)
    builder_test.go:252: 
                Error Trace:    /Users/daniilankusin/monomer/builder/builder_test.go:252
                Error:          Received unexpected error:
                                make block: convert block to Ethereum representation: adapt txs: adapt cosmos deposit tx: unmarshal binary tx: transaction type not supported
                Test:           TestRollback
```
I support your point, I don't think we can leave `break` in there
I will try ti fix `builder_test.go`

[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply mempool/mempool.go]: upd:
Used testutils to generate txs -> panic due to equality of txs
Fixed panic with randomly generated txs
Errors remain:
```
--- FAIL: TestBuild (0.04s)
    --- FAIL: TestBuild/txs_in_inclusion_list (0.01s)
        /Users/daniilankusin/monomer/builder/builder_test.go:133: 
            	Error Trace:	/Users/daniilankusin/monomer/builder/builder_test.go:133
            	Error:      	Received unexpected error:
            	            	make block: convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal cosmos tx: proto: illegal wireType 6
            	Test:       	TestBuild/txs_in_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool (0.01s)
        /Users/daniilankusin/monomer/builder/builder_test.go:133: 
            	Error Trace:	/Users/daniilankusin/monomer/builder/builder_test.go:133
            	Error:      	Received unexpected error:
            	            	make block: convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal cosmos tx: proto: Tx: illegal tag 0 (wire type 2)
            	Test:       	TestBuild/txs_in_mempool
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list (0.00s)
        /Users/daniilankusin/monomer/builder/builder_test.go:133: 
            	Error Trace:	/Users/daniilankusin/monomer/builder/builder_test.go:133
            	Error:      	Received unexpected error:
            	            	make block: convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal cosmos tx: proto: illegal wireType 6
            	Test:       	TestBuild/txs_in_mempool_and_inclusion_list
    --- FAIL: TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool (0.01s)
        /Users/daniilankusin/monomer/builder/builder_test.go:133: 
            	Error Trace:	/Users/daniilankusin/monomer/builder/builder_test.go:133
            	Error:      	Received unexpected error:
            	            	make block: convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal cosmos tx: proto: illegal wireType 6
            	Test:       	TestBuild/txs_in_mempool_and_inclusion_list_with_NoTxPool
```

File: builder/builder.go [hunk 3]
```diff
@@ -104,7 +104,7 @@ type Payload struct {
 }
 
 func (b *Builder) Build(ctx context.Context, payload *Payload) (*monomer.Block, error) {
-	txs := slices.Clone(payload.InjectedTransactions) // Shallow clone is ok, we just don't want to modify the slice itself.
+	cosmosTxs := slices.Clone(payload.InjectedTransactions) // Shallow clone is ok, we just don't want to modify the slice itself.
```
Comment (joshklop) [inline builder/builder.go:107]: It's generally best to avoid renaming unnecessarily. It creates a large diff that's difficult to review.

**Review by joshklop — CHANGES_REQUESTED** [review]
The diff is getting quite large, and its a bit challenging to separate the refactors from the actual changes.

Can you please split this into two PRs: one for the adapters changes, and one for the deposit tx changes? We can focus on the adapters changes for now.

**Comment by AnkushinDaniil** [comment]
> Can you please split this into two PRs: one for the adapters changes, and one for the deposit tx changes? We can focus on the adapters changes for now.
I'm going to minimize changes, but it's impossible to split adapters and depTxs checks because separately it doesn't work and doesn't pass tests.

**Comment by AnkushinDaniil** [comment]
Fixed rollback in builder, now error in localdb:
```
--- FAIL: TestBlockAndHeader (0.00s)
    localdb_test.go:23: 
                Error Trace:    /Users/daniilankusin/monomer/monomerdb/localdb/localdb_test.go:23
                Error:          Received unexpected error:
                                convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal binary tx: transaction type not supported
                Test:           TestBlockAndHeader
--- FAIL: TestRollback (0.00s)
    localdb_test.go:111: 
                Error Trace:    /Users/daniilankusin/monomer/monomerdb/localdb/localdb_test.go:111
                Error:          Received unexpected error:
                                convert block to Ethereum representation: adapt txs: cosmos deposit: unmarshal binary tx: transaction type not supported
                Test:           TestRollback
FAIL
FAIL    github.com/polymerdao/monomer/monomerdb/localdb 3.151s
```

**Comment by AnkushinDaniil** [comment]
# Summary of Changes

1. Mempool deposit transaction check:

   - Split the adapter function into two:
     - One for the core logic
     - One is a check for deposit transactions in the mempool.

1. Bug Fix in Adapter Function:

    - Corrected an error in the adapter function.
    - Specifically, replaced a break statement with an error statement. The original logic was prematurely exiting instead of throwing an error, leading to incorrect behavior in edge cases.

1. Test update:

    - The adapter function fix caused tests to fail
    - Many tests were based on incorrect assumptions or not aligned with logic
    - As a result, I've rewrotten them to reflect correct behavior

## Next Steps

  - Created TODOs to add test cases in future

File: mempool/mempool.go [hunk 4]
```diff
@@ -37,6 +38,13 @@ func (p *Pool) Enqueue(userTxn comettypes.Tx) error {
 	// Unfortunately, comet's DB interface doesn't support it.
 	// Moving to a different DB interface is left for future work.
 
+	// Attempt to adapt the Cosmos transaction to an Ethereum deposit transaction.
+	// If the adaptation results in one or more transactions, it indicates that the
+	// user transaction is a deposit transaction, which is not allowed in the pool.
```
Comment (joshklop) [inline mempool/mempool.go:43]: I think we changed the code but forgot the comment before.
```suggestion
	// If the adaptation succeeds, it indicates that the
	// user transaction is a deposit transaction, which is not allowed in the pool.
```

File: monomerdb/localdb/localdb_test.go [hunk 5]
```diff
@@ -103,28 +114,31 @@ func testHeadBlock(t *testing.T, db *localdb.DB, block *monomer.Block) {
...
 	require.NoError(t, err)
 	require.NoError(t, db.AppendBlock(block))
 	require.NoError(t, db.UpdateLabels(block.Header.Hash, block.Header.Hash, block.Header.Hash))
 
 	block2, err := monomer.MakeBlock(&monomer.Header{
 		Height: 2,
-	}, bfttypes.ToTxs(testapp.ToTxs(t, map[string]string{
-		"k2": "v2",
-	})))
+	}, adaptedTxs)
```
Comment (joshklop) [inline monomerdb/localdb/localdb_test.go:135]: We want to use different txs to test different possible edge cases.

File: x/rollup/types/adapters.go [hunk 6]
```diff
@@ -96,22 +107,18 @@ func AdaptCosmosTxsToEthTxs(cosmosTxs bfttypes.Txs) (ethtypes.Transactions, erro
...
 			return nil, errors.New("MsgL1Tx contains non-deposit tx")
 		}
 		txs = append(txs, &tx)
 	}
 
-	// Pack Cosmos txs into Ethereum txs.
-	for _, txBytes := range txsBytes[1:] {
-		txs = append(txs, AdaptNonDepositCosmosTxToEthTx(txBytes))
-	}
-
```
Comment (joshklop) [inline x/rollup/types/adapters.go:114]: why is this deleted?

**Review by joshklop — CHANGES_REQUESTED** [review]

**Review by joshklop — CHANGES_REQUESTED** [review]

File: builder/builder_test.go [hunk 7]
```diff
@@ -37,54 +43,73 @@ func (*queryAll) String() string {
 }
 
 func TestBuild(t *testing.T) {
-	tests := map[string]struct {
-		inclusionList map[string]string
-		mempool       map[string]string
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply builder/builder_test.go]: Reverted
